core/domain/feature_gating_services_test.py

```diff
@@ -0,0 +1,217 @@
+            self.assertEqual(context.client_type, 'Android')
+            self.assertEqual(context.browser_type, None)
+            self.assertEqual(context.app_version, '1.0.0')
+            self.assertEqual(context.user_locale, 'en')
+
+    def test_get_all_feature_flag_setting_dicts_returns_correct_dicts(self):
+        expected_dicts = {
+            self.param_1.name: self.param_1.to_dict(),
+            self.param_2.name: self.param_2.to_dict(),
+        }
+        self.assertEqual(
+            feature_services.get_all_feature_flag_setting_dicts(),
+            expected_dicts)
+
+    def test_get_all_feature_flag_values_in_dev_returns_correct_values(self):
+        with self.swap(constants, 'DEV_MODE', True):
+            context = feature_services.create_evaluation_context_for_client({
+                'client_type': 'Android',
+                'browser_type': None,
+                'app_version': '1.0.0',
+                'user_locale': 'en',
+            })
+            self.assertEqual(
+                feature_services.get_all_feature_flag_values_for_context(
+                    context),
+                {
+                    self.param_1.name: False,
+                    self.param_2.name: True,
+                }
+            )
+
+    def test_get_all_feature_flag_values_in_prod_returns_correct_values(self):
+        with self.swap(constants, 'DEV_MODE', False):
+            context = feature_services.create_evaluation_context_for_client({
+                'client_type': 'Android',
+                'browser_type': None,
+                'app_version': '1.0.0',
+                'user_locale': 'en',
+            })
+            self.assertEqual(
+                feature_services.get_all_feature_flag_values_for_context(
+                    context),
+                {
+                    self.param_1.name: False,
+                    self.param_2.name: False,
+                }
+            )
+
+    def test_get_feature_flag_value_in_dev_env_returns_correct_values(self):
+        with self.swap(constants, 'DEV_MODE', True):
+            self.assertFalse(
+                feature_services.get_feature_flag_value(self.param_1.name))
+            self.assertTrue(
+                feature_services.get_feature_flag_value(self.param_2.name))
+
+    def test_get_feature_flag_value_in_prod_env_returns_correct_values(self):
+        with self.swap(constants, 'DEV_MODE', False):
+            self.assertFalse(
+                feature_services.get_feature_flag_value(self.param_1.name))
+            self.assertFalse(
+                feature_services.get_feature_flag_value(self.param_2.name))
+
+    def test_get_feature_flag_values_with_invalid_name_failure(self):
+        with self.assertRaisesRegexp(Exception, 'not exist'):
+            feature_services.get_feature_flag_value(
+                'feature_that_does_not_exist')
+
+    def test_update_feature_flag_rules_successfully_updates_rules(self):
```

Comment:
Suggest also adding a test for the case where the dev feature flag is dev targeting, enables for prod, and is in prod mode (would expect the flag to be false since it's not yet enabled for prod despite its conditions targeting prod). This directly verifies that the stages are correctly factored in during evaluation time.

Reply:
I'm not actually following--that test doesn't seem specific to this situation (ensuring that an incompatibility between the target stage & filtered stages results in a validation error). Can you add an explicit test for that situation?

In the essence of keeping this PR chain unblocked, though, I suggest adding that in a subsequent PR that's not part of this chain.

Reply:
Got it, I will make another PR following this one, then the others (controllers, etc.) won't be blocked.

Reply:
Thanks!